### PR TITLE
[RLlib] Cleanup examples folder vol. 23: Add example script for custom metrics on `EnvRunners` (using `MetricsLogger` API).

### PR DESCRIPTION
**Pull request by sven1977**
Cleanup examples folder vol. 23: Add example script for custom metrics on `EnvRunners` (using `MetricsLogger` API).
* Activated in CI
* Example creates 2D heatmap for pacman per episode, logs a custom max. and mean metric (per episode over a sliding window), and the number of lives as EMA-smoothed.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/examples/metrics/custom_metrics_in_env_runners.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    At each episode timestep, the current pacman (y/x)-position is determined and added
+    to the episode's temporary storage. At the end of an episode, a simple 2D heatmap
+    is created from this data and the heatmap is logged to the MetricsLogger (to be
+    viewed in WandB).
+    - the max distance travelled by MsPacman per episode, then averaging these max
+    values over a window of size=100.
+    - the mean distance travelled by MsPacman per episode (over an infinite window).
+    - the number of lifes of MsPacman EMA-smoothed over time.
+
+    This callback can be setup to only log stats on certain EnvRunner indices through
```
Comment (simonsays1980) [inline rllib/examples/metrics/custom_metrics_in_env_runners.py:108]: What happens if an `EnvRunner` crashes?

File: rllib/examples/metrics/custom_metrics_in_env_runners.py [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+when using the new API stack through
+`config.api_stack(_enable_rl_module_and_learner=True,
+_enable_env_runner_and_connector_v2=True)`),
+and which offers a unified API to log individual values per iteration, per episode
+timestep, per episode (as a whole), per loss call, etc..
+`MetricsLogger` objects are available in all custom API code, for example inside your
+custom `Algorithm.training_step()` methods, custom loss functions, custom callbacks,
+and custom EnvRunners.
+
+This example:
```
Comment (simonsays1980) [inline rllib/examples/metrics/custom_metrics_in_env_runners.py:13]: Awesome example!! This shows a lot how to use the `MetricsLogger`. 

To increase complexity, we could:
1. Run this only in evaluation and run evaluation each `n`th training step.
2. Reset some metrics in between.
3. Log two group of metrics (for two groups of env-runners).

File: rllib/examples/metrics/custom_metrics_in_env_runners.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+episodes.
+- The average per-episode distance travelled by Ms Pacman over a sliding window of 200
+episodes.
+- The EMA-smoothed number of lives of Ms Pacman at each timestep (across all episodes).
+
+
+How to run this script
+----------------------
+`python [script file name].py --enable-new-api-stack --wandb-key [your WandB key]
+--wandb-projecy [some project name]`
```
Comment (simonsays1980) [inline rllib/examples/metrics/custom_metrics_in_env_runners.py:46]: "--wandb-projecy" -> "--wandb-project"

Comment (simonsays1980) [reply rllib/examples/metrics/custom_metrics_in_env_runners.py:46]: The same options occur below in "For logging ..."

File: rllib/examples/metrics/custom_metrics_in_env_runners.py [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+            window=200,
+        )
+
+        # Log the number of lifes (as EMA-smoothed; no window).
+        metrics_logger.log_value(
+            "pacman_lifes",
+            episode.get_infos(-1)["lives"],
+            reduce="mean",  # <- default (must be "mean" for EMA smothing)
+            ema_coeff=0.01,  # <- default EMA coefficient (`window` must be None)
+        )
```
Comment (simonsays1980) [inline rllib/examples/metrics/custom_metrics_in_env_runners.py:263]: Maybe add a comment in regard to `metrics_logger.reduce` when to use it and why not here.

File: rllib/utils/metrics/metrics_logger.py [hunk 5]
```diff
@@ -238,20 +238,39 @@ def reduce(
...
+                stats_to_return = self._get_key(key).copy()
+                self._set_key(
+                    key, tree.map_structure_with_path(_reduce, stats_to_return)
+                )
+            else:
+                stats_to_return = self.stats.copy()
+                self.stats = tree.map_structure_with_path(_reduce, stats_to_return)
+        # Provide proper error message if reduction fails due to bad data.
+        except Exception as e:
+            raise ValueError(
```
Comment (simonsays1980) [inline rllib/utils/metrics/metrics_logger.py:267]: Here I wonder what happens when a user resets in between two training steps a key.

**Review by simonsays1980 — APPROVED** [review]
LGTM. Awesome example. Documenting the `MetricsLogger´ and showing users how to use it correctly is so important. I think still that this is one of the complexest parts of `RLlib` and gave some suggestions for further cases.

File: rllib/examples/metrics/custom_metrics_in_env_runners.py (hunk 1 above)
Comment (sven1977) [reply rllib/examples/metrics/custom_metrics_in_env_runners.py:108]: Good question. The index first goes out of commission (other EnvRunners will NOT change their indices b/c of another EnvRunner's crash), but only until the actor is automatically restarted. The latter only happens if `config.recreate_failed_env_runners=True`, of course.

File: rllib/examples/metrics/custom_metrics_in_env_runners.py (hunk 2 above)
Comment (sven1977) [reply rllib/examples/metrics/custom_metrics_in_env_runners.py:13]: Good point! There are two more PRs in flight: MetricsLogger on algorithm.training_step and MetricsLogger inside loss function. A third one could be: MetricsLogger only on eval env runners 🙌

File: rllib/utils/metrics/metrics_logger.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/metrics/metrics_logger.py:267]: Great question. Users should - ideally - never use this API and let RLlib determine, when to call `.reduce()` on a MetricsLogger object. They should only call `.peek()` to see individual reduced stats at the moment w/o actually altering/reducing the underlying stats object.

Comment (sven1977) [reply rllib/utils/metrics/metrics_logger.py:267]: Maybe I need to make this more clear in the docstring ...

File: rllib/examples/metrics/custom_metrics_in_env_runners.py (hunk 3 above)
Comment (sven1977) [reply rllib/examples/metrics/custom_metrics_in_env_runners.py:46]: Yeah, that's ok. The statement below is the generic enable-logging statement, the one above are the recommended args for this particular script.

File: rllib/utils/metrics/metrics_logger.py (hunk 5 above)
Comment (sven1977) [reply rllib/utils/metrics/metrics_logger.py:267]: done
